src/python/pants/engine/internals/graph.py

```diff
@@ -174,20 +174,90 @@ async def find_owners(owners_request: OwnersRequest) -> Owners:
 
     # Walk up the buildroot looking for targets that would conceivably claim changed sources.
     candidate_specs = tuple(AscendantAddresses(directory=d) for d in dirs_set)
-    candidate_targets = await Get(Targets, AddressSpecs(candidate_specs))
+    candidate_explicit_targets = await Get(Targets, AddressSpecs(candidate_specs))
+
+    # Generate subtargets. This gives us file-level precision, where each target only owns one file.
+    # We may not end up using every generated subtarget, but we need to consider each of them. We
+    # also consider the original owning targets so that we can handle deleted files that would have
+    # belonged to those targets.
+    candidate_explicit_target_sources = await MultiGet(
+        Get(HydratedSources, HydrateSourcesRequest(tgt.get(Sources)))
+        for tgt in candidate_explicit_targets
+    )
+    candidate_targets: OrderedSet[Target] = OrderedSet(candidate_explicit_targets)
+    for explicit_tgt, explicit_tgt_sources in zip(
+        candidate_explicit_targets, candidate_explicit_target_sources
+    ):
+        candidate_targets.update(
+            generate_subtarget(explicit_tgt, full_file_name=f)
+            for f in explicit_tgt_sources.snapshot.files
+        )
```

Comment:
Rather than generating all of these before beginning the matching, it seems like you could maybe do it lazily inside the loop when you hit the "Else, we know it's a generated subtarget" condition, because as it stands there will always be an owning "explicit" target.

Reply:
Much better. Thanks for the suggestion!